Add support for highlight selection and onAction to TableView and ListView

devongovett opened the pull request:
Closes #1638. Closes #1905. Closes #2013. Closes #2018. Relates to: #2266.

This implements two features across all collection components (currently TableView and ListView):

* **Highlight selection** – this allows checkboxes to be hidden, and a highlight style to be applied to the row to indicate selection instead. It also changes the selection behavior, so that clicking an item replaces the entire selection with that item rather than toggling it. Multiple items can be selected using modifier keys like Ctrl/Cmd and Shift. On touch, highlight selection behaves like toggle because there are no modifiers available. This can be enabled using the `selectionBehavior="replace"` prop (TBD).
* **onAction** – this prop allows a separate action to be defined for rows in addition to selection. For mouse, this is double click. For keyboard, it's the Enter key. On touch, the action becomes the primary tap interaction, and a long press enters into selection mode. Deselecting all items exits selection mode. The `useLongPress` hook has been added to `@react-aria/interactions` to implement this.

This also required moving the `usePress` call into `useSelectableItem` rather than calling it from each usage site, so there are many changes related to that.

majornista commented:
@devongovett 
1. We need to have a keyboard and screen reader accessible way to select multiple non-contiguous rows when using `selectionBehavior="replace"`. We seem to be able to move row focus without changing the selection using Control+Alt+ArrowUp/Down.  This matches the behavior for non-contiguous selection in Windows Explorer.  However, unlike in Windows Explorer Control+Alt+Space does not toggle selection of the focused row.
2. Depending on the AT, and this applied to the other selectionBehaviors as well, it can be very difficult to distinguish a selected row from a not selected row, and when the selected state changes. This is true for TableView and ListView.
  - VoiceOver seems to operate under the assumption that a row that has focus is also selected, so it doesn't announce the selected state when a row receives focus.
  - JAWS also seems to operate under the assumption that a row that has focus is also selected and does not announce the selected state when the row receiving focus is selected. 
  - VoiceOver and JAWS do not announce selection change at all following a selection event. NVDA will announce when focus is on the row itself, but not if focus is on a cell. 
3. Should Esc still deselect all rows with `selectionBehavior="replace"`.
4. In ListView, [with selectionMode="none" and onAction](https://reactspectrum.blob.core.windows.net/reactspectrum/cf68261f866424efa393715b888482231512e526/storybook-17/index.html?path=/story/listview--selectionmode-none-onaction), the onAction event does not fire for Enter or Space, but fires onPress.
5. I'm concerned that on there is be no way to notify a user that a row can be selected or deselected, particularly on mobile.
6. Related to Grid keyboard navigation when there are focusable interactive elements within the ListView item or TableView:  NVDA switches from forms mode back to browse mode when focus moves to an item that does not require forms mode, like a link or a button in [this example](https://reactspectrum.blob.core.windows.net/reactspectrum/cf68261f866424efa393715b888482231512e526/storybook-17/index.html?path=/story/listview--actions-actionmenu). This means that arrow key navigation between the buttons in the item and up/down to navigate between items in the ListView will no longer work. Also, because we set tabIndex="-1" on both the row and the containing grid when focus is within the grid, it's difficult to to get back to the containing row to get NVDA back into forms mode. I have discovered that if you press Enter with NVDA browse mode focused on static content within the item, like "row 1 description for row 1", NVDA will switch back into forms mode and focus the row, but this behavior is a somewhat flaky.

devongovett commented:
> However, unlike in Windows Explorer Control+Alt+Space does not toggle selection of the focused row.

hmm, are you sure? It does for me...

> VoiceOver and JAWS do not announce selection change at all following a selection event

Yeah, we have live region announcements for this in the previous selectionBehavior, but I turned it off for selectionBehavior=replace because it was live announcing every time you moved focus (because selection follows focus). Was going to ask what we should do about that...

> Should Esc still deselect all rows with selectionBehavior="replace".

It does I believe. Do you think it shouldn't? If so, then how would you deselect all rows?

> In ListView, with selectionMode="none" and onAction, the onAction event does not fire for Enter or Space, but fires onPress.

Seems to be a bug. This works in TableView.

majornista commented:
> > However, unlike in Windows Explorer Control+Alt+Space does not toggle selection of the focused row.
> 
> hmm, are you sure? It does for me...

I'm pretty sure Ctrl+Alt+Space does not toggle selection in this example: https://reactspectrum.blob.core.windows.net/reactspectrum/cf68261f866424efa393715b888482231512e526/storybook-17/index.html?path=/story/listview--selectionbehavior-replace

>
> 
> > VoiceOver and JAWS do not announce selection change at all following a selection event
> 
> Yeah, we have live region announcements for this in the previous selectionBehavior, but I turned it off for selectionBehavior=replace because it was live announcing every time you moved focus (because selection follows focus). Was going to ask what we should do about that...

Perhaps with selectionBehavior="replace", we should only announce when multiple selection changes, as with Shift+Arrow or Ctrl+Alt+Space.

> 
> > Should Esc still deselect all rows with selectionBehavior="replace".
> 
> It does I believe. Do you think it shouldn't? If so, then how would you deselect all rows?

I think it's ok, just a bit odd with the forced selection on focus, to be able to have no selection.

> 
> > In ListView, with selectionMode="none" and onAction, the onAction event does not fire for Enter or Space, but fires onPress.
> 
> Seems to be a bug. This works in TableView.

majornista commented:
> > I'm pretty sure Ctrl+Alt+Space does not toggle selection in this example
> 
> hmm weird it does for me in Chrome and Safari at least. What browser or AT are you using?

No AT, just testing using keyboard. With VoiceOver for macOS, Ctrl+Alt+Space will toggle selection, because VoiceOver handles the event and triggers a click at the center of the element focused with the VoiceOver cursor.

majornista commented:
8. Related to my comment #6 above, action buttons and links within and item should be accessible using the Tab key for when focus is on or within the item containing the buttons or links. The left-right arrow key navigation to access these interactive controls is unexpected for keyboard users and with NVDA will switch the user out of forms mode, breaking arrow key navigation of the ListView, TableView or GridView. cc  @jnurthen

devongovett commented:
Replying to your issues:

1. Fixed. Ended up making this different depending on the platform. On Mac, it's Alt + ArrowUp/ArrowDown/Space. On Windows, it's Ctrl instead. Windows Explorer uses Ctrl (Alt is not necessary) so we match there. In addition, on Mac, Ctrl + ArrowUp/ArrowDown has a system wide binding, as does Ctrl + Alt + Space. On Windows, Alt + Space has a system wide binding (so does Ubuntu Linux). So the only way is to make it specific to the platform.
2. This doesn't seem specific to this PR. We do add `aria-selected="false"` on selectable but not currently selected rows, so AT could technically read this. Not sure how else we would indicate that a row is selectable. We do have the live region announcements on selection changes though.
3. I think Escape is useful as a way to remove the entire selection?
4. Fixed.
5. Same as 2? On mobile it will be even harder though. VO on iOS does not announce aria-labelledby or aria-describedby on grid cells, and you cannot navigate to rows at all.
6. Insert + Space toggles forms mode in NVDA, and I noticed that once you do that it works. Not really sure what we could do about this. Arrow key navigation between cells in a grid seems like a desired feature. Again, doesn't seem related to the changes in this PR.
7. Fixed.
8. This doesn't really have to do with multiple focusable elements within a cell at all though. The same NVDA issue also happens when there is only a single focusable element per cell.

majornista commented:
For ListView, we seem to be missing the blue border-left .is-focused style that is present for ListBox and TableView. We need it to be able to distinguish between `.is-selected` and `.is-selected.is-focused`.

LFDanLu commented:
Just a fyi for others reviewing the PR. 
@majornista with regard to the focused + selected style, was doing some debugging locally and just realized that this didn't have the ListView focus ring changes from main. This is what it looks like now:
![image](https://user-images.githubusercontent.com/9661127/135664772-3b4696fb-40bd-4e65-996b-e8eb22792ac8.png)

There will probably need to be some updates to the focus ring styles vs the selected styles to get rid of the extra stacking top/bottom border + something broke with the selection background color (note the gray and the blue for the same kind of selected row):
![image](https://user-images.githubusercontent.com/9661127/135665557-71d3143b-9039-455e-8259-0a19e03be56c.png)

LFDanLu commented on packages/@react-spectrum/list/src/ListViewItem.tsx:
```diff
@@ -71,7 +76,8 @@ export function ListViewItem(props) {
 
-  let showCheckbox = state.selectionManager.selectionMode !== 'none';
+  let showCheckbox = state.selectionManager.selectionMode !== 'none' && state.selectionManager.selectionBehavior === 'toggle';
+  let isSelected = state.selectionManager.isSelected(item.key);
```
Something broke with the merge from master, looks like the selectedKey isn't being triggered on row press, digging into why

LFDanLu commented on packages/@react-stately/selection/src/SelectionManager.ts:
```diff
@@ -412,4 +438,17 @@ export class SelectionManager implements MultipleSelectionManager {
+    }
+
+    let item = this.collection.getItem(key);
+    if (!item || (item.type === 'cell' && !this.allowsCellSelection)) {
```
Looks like this is the thing blocking ListView item selection after the merge from main. The changes from main (aka the stuff from https://github.com/adobe/react-spectrum/pull/2325/files) made it so that we rely on the press handlers on the cell to toggle selection but this will block the handlers from making it to the cell via the changes in useSelectableItem

LFDanLu replied:
not entirely sure why the ListView row press handlers aren't getting the press event if the ListView cell doesn't have it's press handlers returned

LFDanLu reviewed (commented):
early review submit since master merge broke selection in ListView when clicking/hitting enter on the row. Doesn't happen for TableView, looking into the difference between the two

LFDanLu commented on packages/@react-aria/selection/src/useSelectableItem.ts:
```diff
@@ -66,10 +76,18 @@ export function useSelectableItem(options: SelectableItemOptions): SelectableIte
   } = options;
 
-  let onSelect = (e: PressEvent | PointerEvent) => manager.select(key, e);
+  let onSelect = (e: PressEvent | LongPressEvent | PointerEvent) => {
+    if (e.pointerType === 'keyboard' && isNonContiguousSelectionModifier(e)) {
+      manager.toggleSelection(key);
+    } else {
+      manager.select(key, e);
+    }
```
In the build prior to the master merge and in my local that has the master merge + some rough hacks to get selection working again, I can't seem to toggle selection in the ListView via Enter on an already toggled item. Hitting enter without the option+cmd key will enter the `manager.select` case while hitting enter with the option+cmd key works when toggling selection ON but not off (doesn't even seem to trigger onSelect).

Hitting Enter then Enter + Option + CMD is the only way I can get it to untoggle properly

LFDanLu commented on packages/@react-spectrum/table/src/TableView.tsx:
```diff
@@ -533,9 +544,10 @@ function TableRow({item, children, ...otherProps}) {
           {
             'is-active': isPressed,
```
The TableView "is-active" styles seem to persist even when the row is no longer pressed.
Specific reproduction:
1. Using option+command+Enter, select a bunch of rows in the "selectionBehavior: replace" story
2. Press up/down arrow which should deselect all rows except the one you are now focusing 
3. Note that the previously selected rows are now grey

![image](https://user-images.githubusercontent.com/9661127/135675874-37011638-e941-40c3-9eff-a86d5b156dbf.png)
Happens for ListView as well

LFDanLu reviewed (commented):
Couple more things from TableView/ListView testing and a quick code scan. Haven't tried mobile or accessibility yet, thats next

LFDanLu commented on packages/@react-aria/selection/stories/useSelectableList.stories.tsx:
```diff
@@ -131,4 +132,54 @@ storiesOf('useSelectableList', module)
+  .add(
+    'single select, allow empty, select on focus',
+    () => (
+      <List selectionMode="single">
+        <Item>Paco de Lucia</Item>
+        <Item>Vicente Amigo</Item>
+        <Item>Gerardo Nunez</Item>
+      </List>
```
This allows you to select multiple options via moving the focus via option+arrow keys and option+Enter/Space. Same with the single select story below

LFDanLu commented on packages/@react-spectrum/table/stories/Table.stories.tsx:
```diff
@@ -280,6 +280,57 @@ storiesOf('TableView', module)
+   .add(
+    'selectionMode: none, onAction',
+    () => (
+      <TableView aria-label="TableView with dynamic contents" width={500} height={400} onSelectionChange={s => onSelectionChange([...s])} onAction={action('onAction')}>
+        <TableHeader columns={columns}>
+          {column => <Column>{column.name}</Column>}
+        </TableHeader>
+        <TableBody items={items}>
+          {item =>
+            (<Row key={item.foo}>
+              {key => <Cell>{item[key]}</Cell>}
+            </Row>)
+          }
+        </TableBody>
+      </TableView>
```
Option+Enter is causing onSelectionChange to fire here

LFDanLu reviewed (changes requested):
Menus also look weird after the merge with main:
![image](https://user-images.githubusercontent.com/9661127/135688233-f5d0b806-6660-41eb-bf4b-8b9fd9a77f34.png)

dannify commented:
Some follow up tickets will be added for this.
